CaloCluster/src/CaloClusterFast_module.cc

```diff
@@ -1,322 +1,197 @@
-                   {
-                       int index = int(t0/digiSampling_) + nCount - winOffsetT0_;
-		      
-                       hitList_[index].push_back(FastHit(crId,index,deque_.front()));
-                       if (deque_.front()> minSeedAmp_) seeds_.emplace_back(&(hitList_[index].back()));
-                   }
-
-                   if (*tail == deque_.front()) deque_.pop_front();
-                   ++tail;
-                }
-                ++nCount;
-           } 
+  //----------------------------------------------------------------------------------------------------------
+  void CaloClusterFast::makeClusters(CaloClusterCollection& caloClusters, const art::Handle<CaloHitCollection>& caloHitsHandle)
+  {
+      const Calorimeter& cal = *(GeomHandle<Calorimeter>());
+      const CaloHitCollection& caloHits(*caloHitsHandle);
+      if (caloHits.empty()) return;
+
+      std::vector<int> hits(caloHits.size());
+      std::iota(hits.begin(),hits.end(),0);
+      
+      auto functorTime = [&caloHits](int a, int b) {return caloHits[a].time() < caloHits[b].time();};
+      std::sort(hits.begin(),hits.end(),functorTime);
+      for (auto val : hits) if (caloHits[val].energyDep() < EnoiseCut_) val = -1;
+
+      auto iterSeed = hits.begin();
+      while (iterSeed != hits.end())
+      {
+          //find the first hit above the energy threshold, and the last hit within the required time window
+          const CaloHit& hitSeed = caloHits[*iterSeed];
+          if (*iterSeed==-1 || hitSeed.energyDep()< EminSeed_) {++iterSeed; continue;}
+          double timeStart = hitSeed.time();
+
+          //find the range around the seed time to search for other hits to form clusters
+          auto iterStart(iterSeed), iterStop(iterSeed);
+          while (iterStop  != hits.end()   && (*iterStop==-1  || caloHits[*iterStop].time() - timeStart < deltaTime_))  ++iterStop;
+          while (iterStart != hits.begin() && (*iterStart==-1 || timeStart - caloHits[*iterStart].time() < deltaTime_)) --iterStart;
+          ++iterStart; 
+
+          //start the clustering algorithm for the hits between iStart and iStop
+          std::queue<int> crystalToVisit;
+          std::vector<bool> isVisited(cal.nCrystal()); 
+
+          //put the first hit in the cluster list
+          std::vector<int> clusterList{*iterSeed};
+          int seedId = caloHits[*iterSeed].crystalID();
+          crystalToVisit.push(seedId);
+          *iterSeed=-1;
+
+          // loop until all seeds are processed
+          while (!crystalToVisit.empty())
+          {            
+              int visitId = crystalToVisit.front();
+              isVisited[visitId]=true;
+
+              std::vector<int>  neighborsId = cal.crystal(visitId).neighbors();
+              if (extendSearch_) std::copy(cal.nextNeighbors(visitId).begin(), cal.nextNeighbors(visitId).end(), std::back_inserter(neighborsId));
+
+              for (auto& iId : neighborsId)
```

Comment:
I think this could be auto const&

Reply:
Almost, this should obviously be const auto& !!!!!!